cli/src/lib.rs

```diff
@@ -187,11 +188,23 @@ where
         );
 
         let query_validator = Arc::new(query_validator);
+
+        // TODO #2292: Very dirty, need to do something with it
+        let world_wsv = if std::any::TypeId::of::<W>() == std::any::TypeId::of::<World>() {
```

Comment:
why are you're getting rid of the generic here?

Reply:
Why do we have `WorldTrait`/`KuraTrait`/`SumeragiTrait`/etc.? I thought this was to be able to swap out different parts of Iroha with different implementations, i.e. to give the blockchain user the ability to customize their blockchain

Reply:
> why are you're getting rid of the generic here?

Because it's impossible to work with generics and deserialization at the same time. And after removing in one place (or giving the ability to work only on some specific types implementing trait) it's unwinding through all the project. I'd like to remove `WorldTrait` at all. See also #2292 


> Why do we have WorldTrait/KuraTrait/SumeragiTrait/etc.? I thought this was to be able to swap out different parts of Iroha with different implementations, i.e. to give the blockchain user the ability to customize their blockchain

I can't say about others, but as I understand, `WorldTrait` was introduced just for mocking purposes...

Reply:
If it's just mocking, I'd try to get rid of all these traits and do it through some form of dependency injection. But is it just for mocking?

Reply:
@appetrosyan , need your answers here, please

Reply:
I don't think that we at any point will have more than one legitimate `World` implementation. They could be versioned, yes, but that system doesn't require generics and traits. 

Dependency injection is probably the right approach, but only if we decide to postpone this PR until after the LTS.

Reply:
ok, and we should assume 3rd parties will not swap in their own implementations of parts of Iroha? What about other traits such as `SumeragiTrait` or `KuraTrait`? Should we aim to get rid of those. I would if their only purpose is mocking, or at least use `#[cfg(test)]` to exclude them from the public API

Reply:
At least `SumeragiTrait` could be related to the [Consensus API](https://wiki.hyperledger.org/display/iroha/Consensus+API) idea in the future. In that case, we would have to extract the logics common to consensus to the trait methods

Reply:
Since Iroha is Open Source, I think that the `ConsensusAPI` doesn't necessarily require a single `SumeragiTrait`, but rather having the flexibility to easily swap one implementation for another. As such, we could retain the trait, but avoid the usage of generics. 

This can change when specialisation is stable.

Reply:
I discussed this with @Arjentix . I think we can provide builder methods to `Iroha` such as `Iroha::with_sumeragi`, etc. and use current implementation as the default if the builder method isn't called

Reply:
That works too.